gateway-sp-comms/src/ereport.rs

```diff
@@ -0,0 +1,831 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2025 Oxide Computer Company
+
+use crate::error::CommunicationError;
+use crate::error::EreportError;
+use crate::shared_socket;
+use crate::single_sp;
+use crate::SpRetryConfig;
+use async_trait::async_trait;
+pub use gateway_messages::ereport::Ena;
+use gateway_messages::ereport::EreportRequest;
+use gateway_messages::ereport::EreportResponseHeader;
+use gateway_messages::ereport::RequestV0;
+use gateway_messages::ereport::ResponseHeaderV0;
+pub use gateway_messages::ereport::RestartId;
+use serde::Deserialize;
+use serde_cbor::Value as CborValue;
+use serde_json::Value as JsonValue;
+use slog::debug;
+use slog::error;
+use slog::trace;
+use slog::warn;
+use slog_error_chain::SlogInlineError;
+use std::collections::BTreeMap;
+use std::num::NonZeroU8;
+use thiserror::Error;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+
+pub const SP_PORT: u16 = 0xDEAD;
+pub const MGS_PORT: u16 = 0xDEAF;
+
+#[derive(Debug, Default)]
+pub struct EreportHandler {}
+
+#[derive(Debug)]
+pub struct EreportTranche {
+    pub restart_id: RestartId,
+    pub ereports: Vec<Ereport>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Ereport {
+    pub ena: Ena,
+    pub data: MetadataMap,
+}
+
+pub(crate) struct WorkerRequest {
+    pub(crate) restart_id: RestartId,
+    pub(crate) start_ena: Ena,
+    pub(crate) limit: NonZeroU8,
+    pub(crate) committed_ena: Option<Ena>,
+    pub(crate) rsp_tx: oneshot::Sender<Result<EreportTranche, EreportError>>,
+}
+
+pub(crate) struct Worker<S> {
+    req_rx: mpsc::Receiver<WorkerRequest>,
+    retry_config: SpRetryConfig,
+    socket: S,
+    outbuf: [u8; <EreportRequest as hubpack::SerializedSize>::MAX_SIZE],
+    metadata: Option<MetadataMap>,
+}
+
+type MetadataMap = serde_json::Map<String, JsonValue>;
+
+impl<S> Worker<S>
+where
+    S: single_sp::InnerSocket<Vec<u8>>,
+{
+    pub(crate) fn new(
+        retry_config: SpRetryConfig,
+        socket: S,
+    ) -> (Self, mpsc::Sender<WorkerRequest>) {
+        let (tx, req_rx) = mpsc::channel(8);
+        let this = Self {
+            req_rx,
+            retry_config,
+            socket,
+            outbuf: [0u8;
+                <EreportRequest as hubpack::SerializedSize>::MAX_SIZE],
+            metadata: None,
+        };
+        (this, tx)
+    }
+
+    pub(crate) async fn run(mut self) {
+        while let Some(req) = self.req_rx.recv().await {
+            // If we have just started up, or we were previously unable to
+            // refresh the SP's metadata, try to do so now.
+            if self.metadata.is_none() {
+                trace!(self.log(), "requesting initial SP metadata...");
+                match self.request_ereports(RestartId(0), Ena(0), 0, None).await
+                {
+                    Ok((restart_id, ereports)) => {
+                        debug_assert!(
+                            ereports.is_empty(),
+                            "we asked for limit 0..."
```

Comment:
I'm tempted to give this a pass since it's a `debug_assert!`, but it seems a little fishy: we're asserting something that we have no control over, right (i.e, this assert fails if the remote end is misbehaving)? I'm inclined to say this should be treated as a runtime error instead.

Reply:
The debug assert was really just intended to catch this in testing, but I agree that it is not under MGS' control. We should maybe just return an error to the requester instead, I think you're right about that.